include/common/Driver.hpp

```diff
@@ -555,75 +554,27 @@ class Driver
 		}
 
 		/*Implementatoin of writing last state Y and YDot for each unit operation*/
-
-		for (int i = 0; i <= _sim->model()->maxUnitOperationId(); ++i)
+		std::ostringstream oss;
+		for (int i = 0; i < _writeLastStateUnitId.size(); ++i)
 		{
-			std::ostringstream oss;
+			const unsigned int lastStateUnitId = _writeLastStateUnitId[i];
 			oss.str("");
-			oss << "unit_" << std::setfill('0') << std::setw(3) << std::setprecision(0) << i;
+			oss << "unit_" << std::setfill('0') << std::setw(3) << std::setprecision(0) << lastStateUnitId;
 			writer.pushGroup("solution");
 			writer.pushGroup(oss.str());
-			if (_flag[i] == true)
-			{
-				unsigned int len = 0;
-				double const* const lastY = _sim->getLastSolution(len);
-				double const* const lastYdot = _sim->getLastSolutionDerivative(len);
-
-				const unsigned int* index_stride = _sim->model()->getModelSlice();
-				int count;
-				std::vector<double> dataVecTemp;
-				std::vector<double> lastYdotTemp;
-				count = 0;
-				for (int j = index_stride[i]; j < index_stride[i + 1]; j++)
-				{
-					dataVecTemp.insert(dataVecTemp.begin() + count, lastY[j]);
-					lastYdotTemp.insert(lastYdotTemp.begin() + count, lastYdot[j]);
-					count++;
-				}
-
-				writer.template vector<double>("LAST_STATE_Y", dataVecTemp.size(), dataVecTemp.data());
-				writer.template vector<double>("LAST_STATE_YDOT", lastYdotTemp.size(), lastYdotTemp.data());
-			}
-			writer.popGroup();
-			writer.popGroup();
-		}
-
-		/*if(_writeLastStateUnit)
-		{
+			
 			unsigned int len = 0;
 			double const* const lastY = _sim->getLastSolution(len);
 			double const* const lastYdot = _sim->getLastSolutionDerivative(len);
+			const unsigned int* modelOffsets = _sim->model()->getModelStateOffsets();
 
-			std::vector<int> index_stride = _sim->model()->getModelSlice();
-			int count;
-
+			writer.template vector<double>("LAST_STATE_Y", modelOffsets[lastStateUnitId + 1] - modelOffsets[lastStateUnitId], lastY + modelOffsets[lastStateUnitId]);
+			writer.template vector<double>("LAST_STATE_YDOT", modelOffsets[lastStateUnitId + 1] - modelOffsets[lastStateUnitId], lastYdot + modelOffsets[lastStateUnitId]);
```

Comment:
This also stores the additional inlet DOFs at the beginning of a state vector slice. Is this intentional?